### PR TITLE
sw-cache-expiration unit tests

### PR DESCRIPTION
**Pull request by jeffposnick**
R: @addyosmani @gauntface

Unit tests for `sw-cache-expiration`. There's also a bug fix for an issue I found in the underlying library (yay testing!) and I'm exposing a few additional symbols that could be useful.

The unit tests are fairly in depth, so I'm not sure that I'm going to prioritize writing end-to-end tests. The in-depth-ness also means that they're a little gnarly, so please feel free to push back if it's too unclear what's going on.

### CODE REVIEW

File: packages/sw-cache-expiration/src/lib/plugin.js [hunk 1]
```diff
@@ -303,7 +326,9 @@ class Plugin {
    * @param {string} input.cacheName Name of the cache the Responses belong to.
    * @return {Array<string>} A list of the URLs that were expired.
    */
-  async findExtraEntries({cacheName}) {
+  async findExtraEntries({cacheName} = {}) {
```
Comment (gauntface) [inline packages/sw-cache-expiration/src/lib/plugin.js]: Nit: But this name seems like the opposite of the description: "findExtraEntries" > "Expires entries base on the the maximum cache size."

File: packages/sw-cache-expiration/test/sw/plugin.js [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+    return plugin.getCache({cacheName}).then((firstCache) => {
+      return plugin.getCache({cacheName}).then((secondCache) => {
+        expect(firstCache).to.eql(secondCache);
+      });
+    });
+  });
+
+  it(`should return true when there's no cachedResponse passed to isResponseFresh()`, function() {
+    const plugin = new Plugin({maxAgeSeconds});
+    expect(plugin.isResponseFresh()).to.be.true;
```
Comment (gauntface) [inline packages/sw-cache-expiration/test/sw/plugin.js]: This feels a bit of a weird test in that I would have expected `inResponseFresh()` to throw if there was not response to check for freshness.

Is there a use case for this returning true?

File: packages/sw-cache-expiration/test/sw/plugin.js [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+    const plugin = new Plugin({maxAgeSeconds});
+    const date = new Date(now).toUTCString();
+    const cachedResponse = new Response('', {headers: {date}});
+    expect(plugin.isResponseFresh({cachedResponse, now})).to.be.true;
+  });
+
+  it(`should return false when isResponseFresh() is called and the Date: header in the cachedResponse is not recent`, function() {
+    const plugin = new Plugin({maxAgeSeconds});
+    // This will construct a date that is 1 second past the expiration.
+    const date = new Date(now - ((1 + maxAgeSeconds) * 1000)).toUTCString();
```
Comment (gauntface) [inline packages/sw-cache-expiration/test/sw/plugin.js]: Feel free to ignore this but worth considering:

You could condense some of these tests down into a single unit test (i.e. fresh with time as normal, then advance time and test again, still fresh, then advance past the maximum and ensure not fresh).

Using sinon fake timers you can control the time of the environment: (Example of using them here: https://github.com/gauntface/route/blob/master/packages/device-clock/test/node/device-clock.js#L17 and advance with https://github.com/gauntface/route/blob/master/packages/device-clock/test/node/device-clock.js#L46) Was thoroughly impressed how easy it was to use - no worries if you are happy leaving as is though.

File: packages/sw-cache-expiration/test/sw/plugin.js [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+    const cachedResponse = new Response('', {headers: {date}});
+    expect(plugin.isResponseFresh({cachedResponse, now})).to.be.false;
+  });
+
+  it(`should return cachedResponse when cacheWillMatch() is called and isResponseFresh() is true`, function() {
+    const plugin = new Plugin({maxAgeSeconds});
+    const date = new Date(now).toUTCString();
+    const cachedResponse = new Response('', {headers: {date}});
+    expect(plugin.cacheWillMatch({cachedResponse, now})).to.eql(cachedResponse);
```
Comment (gauntface) [inline packages/sw-cache-expiration/test/sw/plugin.js]: Have you added now as an argument to accommodate testing or is it useful as part of the API?

File: packages/sw-cache-expiration/src/lib/plugin.js (hunk 1 above)
Comment (jeffposnick) [inline packages/sw-cache-expiration/src/lib/plugin.js]: Thanks; they JSDocs were wrong.

File: packages/sw-cache-expiration/test/sw/plugin.js (hunk 2 above)
Comment (jeffposnick) [inline packages/sw-cache-expiration/test/sw/plugin.js]: The underlying method gets called as part of the `RequestWrapper` lifecycle callback `cacheWillMatch`. `cacheWillMatch` might be called with an undefined `Response` passed in if there isn't yet an entry in the cache for a given URL, and we just want to return `true` in that case to work with the semantics that `RequestWrapper` expects.

File: packages/sw-cache-expiration/test/sw/plugin.js (hunk 4 above)
Comment (jeffposnick) [inline packages/sw-cache-expiration/test/sw/plugin.js]: I could imagine someone who's using the underlying module outside of the larger framework wanting that level of control.

File: packages/sw-cache-expiration/src/lib/error-factory.js [hunk 5]
```diff
@@ -0,0 +1,27 @@
+/*
+ Copyright 2016 Google Inc. All Rights Reserved.
```
Comment (jpmedley) [inline packages/sw-cache-expiration/src/lib/error-factory.js]: Should the copyright year be updated?

File: packages/sw-cache-expiration/test/sw/plugin.js (hunk 3 above)
Comment (jeffposnick) [inline packages/sw-cache-expiration/test/sw/plugin.js]: I was thinking of testing that sort of scenario as part of end-to-end tests, but my goal right now is to get some sort of test coverage for all the code, and then go back and cover additional scenarios. So I'll defer on that for now.

File: packages/sw-cache-expiration/src/lib/error-factory.js (hunk 5 above)
Comment (jeffposnick) [inline packages/sw-cache-expiration/src/lib/error-factory.js]: We could, but let's just do that in a separate PR that covers the entire set of modules if we care.
